components/nimbus/src/behavior.rs

```diff
@@ -0,0 +1,177 @@
+#![allow(dead_code)]
+
+use crate::error::{BehaviorError, NimbusError, Result};
+use chrono::{DateTime, Timelike, Utc};
+use std::collections::{HashMap, VecDeque};
+
+#[derive(Clone)]
+enum Interval {
+    MINUTES,
+    HOURS,
+    DAYS,
+    WEEKS,
+    MONTHS,
+    YEARS,
+}
+
+#[derive(Clone)]
+struct IntervalConfig {
+    bucket_count: usize,
+    interval: Interval,
+}
+
+pub struct IntervalData {
+    buckets: VecDeque<u64>,
+    bucket_count: usize,
+    starting_instant: DateTime<Utc>,
+}
+
+impl Default for IntervalData {
+    fn default() -> Self {
+        Self::new(1)
+    }
+}
+
+impl IntervalData {
+    pub fn new(bucket_count: usize) -> IntervalData {
+        let mut data = IntervalData {
+            buckets: VecDeque::new(),
+            bucket_count: bucket_count,
+            starting_instant: Utc::now(),
+        };
+        data.buckets.push_front(0);
+        data
+    }
+
+    fn increment(&mut self) -> Result<()> {
+        match self.buckets.front_mut() {
+            Some(x) => *x += 1,
+            None => {
+                return Err(NimbusError::BehaviorError(BehaviorError::InvalidState(
+                    "Interval buckets cannot be empty".to_string(),
+                )))
+            }
+        };
+        Ok(())
+    }
+
+    fn rotate(&mut self, num_rotations: u64) -> Result<()> {
+        for _ in 1..=num_rotations {
```

Comment:
> And I wonder if appending multiple entries to the "front" of the queue would be more efficient than a loop here?

Can you elaborate on what you mean?

Reply:
After looking at VecDeque docs, I don't see a good way to insert a group of items like this, so disregard this comment. I was mostly interested if we could insert all of the new buckets at the front in a single operation rather than in multiple calls to `push_front`.